buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -767,79 +666,201 @@ public static void appendPrettyHexDump(StringBuilder dump, ByteBuf buf) {
-        } else {
-            dump.append(NEWLINE);
-            dump.append(Long.toHexString(rowStartIndex & 0xFFFFFFFFL | 0x100000000L));
-            dump.setCharAt(dump.length() - 9, '|');
-            dump.append('|');
+        static void appendPrettyHexDump(StringBuilder dump, ByteBuf buf, int offset, int length) {
+            if (isOutOfBounds(offset, length, buf.capacity())) {
+                throw new IndexOutOfBoundsException(
+                        "expected: " + "0 <= offset(" + offset + ") <= offset + length(" + length
+                                                    + ") <= " + "buf.capacity(" + buf.capacity() + ')');
+            }
+            if (length == 0) {
+                return;
+            }
+            dump.append(
+                                     "         +-------------------------------------------------+" +
+                StringUtil.NEWLINE + "         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |" +
+                StringUtil.NEWLINE + "+--------+-------------------------------------------------+----------------+");
+
+            final int startIndex = offset;
+            final int fullRows = length >>> 4;
+            final int remainder = length & 0xF;
+
+            // Dump the rows which have 16 bytes.
+            for (int row = 0; row < fullRows; row ++) {
+                int rowStartIndex = (row << 4) + startIndex;
+
+                // Per-row prefix.
+                appendHexDumpRowPrefix(dump, row, rowStartIndex);
+
+                // Hex dump
+                int rowEndIndex = rowStartIndex + 16;
+                for (int j = rowStartIndex; j < rowEndIndex; j ++) {
+                    dump.append(BYTE2HEX[buf.getUnsignedByte(j)]);
+                }
+                dump.append(" |");
+
+                // ASCII dump
+                for (int j = rowStartIndex; j < rowEndIndex; j ++) {
+                    dump.append(BYTE2CHAR[buf.getUnsignedByte(j)]);
+                }
+                dump.append('|');
+            }
+
+            // Dump the last row which has less than 16 bytes.
+            if (remainder != 0) {
+                int rowStartIndex = (fullRows << 4) + startIndex;
+                appendHexDumpRowPrefix(dump, fullRows, rowStartIndex);
+
+                // Hex dump
+                int rowEndIndex = rowStartIndex + remainder;
+                for (int j = rowStartIndex; j < rowEndIndex; j ++) {
+                    dump.append(BYTE2HEX[buf.getUnsignedByte(j)]);
+                }
+                dump.append(HEXPADDING[remainder]);
+                dump.append(" |");
+
+                // Ascii dump
+                for (int j = rowStartIndex; j < rowEndIndex; j ++) {
+                    dump.append(BYTE2CHAR[buf.getUnsignedByte(j)]);
+                }
+                dump.append(BYTEPADDING[remainder]);
+                dump.append('|');
+            }
+
+            dump.append(StringUtil.NEWLINE +
```

Comment:
consider adding a `static import ...StringUtil.NEWLINE` to avoid having `StringUtil.` on every usage.